File: Fruits.js
/* Fruit constructor creates an object of the prototype then returns it */
function Fruit(Weight) {
    var instance = Object.create(Fruit.prototype);

    instance.weight = Weight;
    /* intialize the parent node to undefined */
    instance.parent = undefined;
    /* intialize the left node to undefined */
    instance.left = undefined;
    /* intialize the right node to undefined */
    instance.right = undefined;

    return instance;
}

Fruit.prototype.getType = function() {
    return "Fruit";
}

Fruit.prototype.Insert = function(fruit) {
    var x = this;
    var y = undefined;

    while(x != undefined) {
        y = x;
        if(fruit.weight < x.weight) {
            x = x.left;
        } else {
            x = x.right;
        }
    }

    if(y == undefined) {
        y = fruit;
    } else if(fruit.weight < y.weight) {
        y.left = fruit;
        fruit.parent = y;
    } else {
        y.right = fruit;
        fruit.parent = y;
    }
}

/* Iterate over all node in the tree and print them in an ascending order */
Fruit.prototype.Iterate = function() {
    var TreeRoot = this;
    function recursiveInorder(root) {
        if(root == undefined) {
            return;
        }
        recursiveInorder(root.left);
        console.log("Fruit Type:", root.getType(), "-", "Weight:", root.weight, ",");
        recursiveInorder(root.right);
    }
    return recursiveInorder(TreeRoot);
}

Fruit.prototype.filterByType = function(Type) {
    // Traverse the tree all the nodes of the Type passed
    function recursiveInorder(root) {
        if(root == undefined) {
            return;
        }
        recursiveInorder(root.left);
        if(root.getType() == Type) {
            console.log("Fruit Type:", root.getType(), "-", "Weight:", root.weight, ",");
        }
        recursiveInorder(root.right);
    }

    return recursiveInorder(this);
}

Fruit.prototype.filterByWeight = function(Weight) {
    // prints all nodes higher than Weight using two helper function passed by another helper function
    higherThan(ceilingNode(this, Weight), successor);
}

Fruit.prototype.magnifyByType = function(Type, Weight) {
    // Traverse the tree and magnify all the nodes of the Type passed
    function recursiveInorder(root) {
        if(root == undefined) {
            return;
        }
        recursiveInorder(root.left);
        if(root.getType() == Type) {
            root.weight += Weight;
        }
        recursiveInorder(root.right);
    }

    return recursiveInorder(this);
}

Fruit.prototype.findLightest = function() {
    var root = this;
    if(root == undefined) {
        return;
    }

    while(root.left != undefined) {
        root = root.left;
    }
    return root;
}

Fruit.prototype.findHeaviest = function () {
    var root = this;
    if(root == undefined) {
        return;
    }
    while(root.right != null) {
        root = root.right;
    }
    return root;
}

/* Helper functions  */
function ceilingNode(root, key) {
    while(root != undefined) {
        if(root.weight == key) {
            break;
        } else if(root.weight > key) {
            root = root.leftt;
        } else {
            root = root.right;
        }
    }
    return root;
}

function successor(node) {
    var p = node.parent;
    if(node.right != undefined) {
        node = node.right;
        while(node.left != undefined) {
            node = node.left;
        }
        return node;
    }

    while(p != undefined && node == p.right) {
        node = p;
        p = p.parent;
    }
    return p;
}

function higherThan(node, func) {
    var n = func(node);
    while(n != undefined) {
        console.log("Fruit Type:", n.getType(), "-", "Weight:", n.weight, ","); 
        n = func(n);
    }
}

/* ........................................................................................................ */
/* Fruits hierarchy */
// Citrus Category
function Citrus(Weight) {
    var instance = new Fruit();
    instance.weight = Weight;
    instance.getType = function() {
        return "Citrus";
    }
    return instance;s
}

function Sweet(Weight) {
    var instance = new Citrus();
    instance.weight = Weight;
    instance.getType = function() {
        return "Sweet ";
    }
    return instance;
}
function Sour(Weight) {
    var instance = new Citrus();
    instance.weight = Weight;
    instance.getType = function() {
        return "Sour ";
    }
    return instance;
}

function Orange(Weight) {
    var instance = new Sweet();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Sweet();
        let parent2 = new Citrus();
        return "Orange ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Tangerine(Weight) {
    var instance = new Sweet();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Sweet();
        let parent2 = new Citrus();
        return "Tangerine ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Mandarin(Weight) {
    var instance = new Sweet();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Sweet();
        let parent2 = new Citrus();
        return "Mandarin ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Lemon(Weight) {
    var instance = new Sour();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Sour();
        let parent2 = new Citrus();
        return "Lemon ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Grapefruit(Weight) {
    var instance = new Sour();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Sour();
        let parent2 = new Citrus();
        return "Grapefruit ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}

// Melons Category
function Melon(Weight) {
    var instance = new Fruit();
    instance.weight = Weight;
    instance.getType = function() {
        return "Melon";
    }
    return instance;
}

function Watermelon(Weight) {
    var instance = new Melon();
    instance.weight = Weight;
    instance.getType = function() {
        let parent = new Melon();
        return "Watermelon ".concat(parent.getType());
    }
    return instance;
}
function Cantaloupe(Weight) {
    var instance = new Melon();
    instance.weight = Weight;
    instance.getType = function() {
        let parent = new Melon();
        return "Cantaloupe ".concat(parent.getType());
    }
    return instance;
}
function Galiamelon(Weight) {
    var instance = new Melon();
    instance.weight = Weight;
    instance.getType = function() {
        let parent = new Melon();
        return "Galiamelon ".concat(parent.getType());
    }
    return instance;
}

// Tropical fruit Category
function Tropical(Weight) {
    var instance = new Fruit();
    instance.weight = Weight;
    instance.getType = function() {
        return "Tropical"
    }
    return instance;
}

function Oval(Weight) {
    var instance = new Tropical();
    instance.weight = Weight;
    instance.getType = function() {
        return "Oval ";
    }
    return instance;
}
function Curved(Weight) {
    var instance = new Tropical();
    instance.weight = Weight;
    instance.getType = function() {
        return "Curved ";
    }
    return instance;
}
function Round(Weight) {
    var instance = new Tropical();
    instance.weight = Weight;
    instance.getType = function() {
        return "Round ";
    }
    return instance;
}

function Avocado(Weight) {
    var instance = new Oval();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Oval();
        let parent2 = new Tropical();
        return "Avocado ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Pineapple(Weight) {
    var instance = new Oval();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Oval();
        let parent2 = new Tropical();
        return "Pineapple ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Banana(Weight) {
    var instance = new Curved();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Curved();
        let parent2 = new Tropical();
        return "Banana ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Coconut(Weight) {
    var instance = new Round();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Round();
        let parent2 = new Tropical();
        return "Coconut ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Dragonfruit(Weight) {
    var instance = new Round();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Round();
        let parent2 = new Tropical();
        return "Dragonfruit ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}

// Berries Category
function Berry(Weight) {
    var instance = new Fruit();
    instance.weight = Weight;
    instance.getType = function() {
        return "Berry";
    }
    return instance;
}

function Tiny(Weight) {
    var instance = new Berry();
    instance.weight = Weight;
    instance.getType = function() {
        return "Tiny ";
    }
    return instance;
}
function Small(Weight) {
    var instance = new Berry();
    instance.weight = Weight;
    instance.getType = function() {
        return "Small ";
    }
    return instance;
}
function Medium(Weight) {
    var instance = new Berry();
    instance.weight = Weight;
    instance.getType = function() {
        return "Medium ";
    }
    return instance;
}

function Elderberry(Weight) {
    var instance = new Tiny();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Tiny();
        let parent2 = new Berry();
        return "ElderBerry ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Blueberry(Weight) {
    var instance = new Small();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Small();
        let parent2 = new Berry();
        return "Blueberry ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Huckleberry(Weight) {
    var instance = new Small();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Small();
        let parent2 = new Berry();
        return "Huckleberry ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Blackberry(Weight) {
    var instance = new Medium();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Medium();
        let parent2 = new Berry();
        return "Blackberry ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}
function Raspberry(Weight) {
    var instance = new Medium();
    instance.weight = Weight;
    instance.getType = function() {
        let parent1 = new Medium();
        let parent2 = new Berry();
        return "Raspberry ".concat(parent1.getType(), parent2.getType());
    }
    return instance;
}

// Stone fruit Category
function Stonefruit(Weight) {
    var instance = new Fruit();
    instance.weight = Weight;
    instance.getType = function() {
        return "Stonefruit";
    }
    return instance;
}

function Plum(Weight) {
    var instance = new Stonefruit();
    instance.weight = Weight;
    instance.getType = function() {
        let parent = new Stonefruit();
        return "Plum ".concat(parent.getType());
    }
    return instance;
}
function Apricot(Weight) {
    var instance = new Stonefruit();
    instance.weight = Weight;
    instance.getType = function() {
        let parent = new Stonefruit();
        return "Apricot ".concat(parent.getType());
    }
    return instance;
}
function Nectarine(Weight) {
    var instance = new Stonefruit();
    instance.weight = Weight;
    instance.getType = function() {
        let parent = new Stonefruit();
        return "Nectarine ".concat(parent.getType());
    }
    return instance;
}
function Peach(Weight) {
    var instance = new Stonefruit();
    instance.weight = Weight;
    instance.getType = function() {
        let parent = new Stonefruit();
        return "Peach ".concat(parent.getType());
    }
    return instance;
}

File: Fruits.scala
trait Fruit {
  var weight :Int =0 
  var left : Fruit = null
  var right : Fruit = null
  
  def setweight(w :Int){
    weight = w
  }
  
  def setleft(left :Fruit){
    this.left = left
  }
  
  def setright(right :Fruit){
    this.right = right
  }
  
  def getname():String
  
}

trait BigFruit extends Fruit{ def getname():String ={ "Big"}}

trait OvalFruit extends Fruit{ def getname():String ={ "Oval"}}

trait TinyFruit extends Fruit{ def getname():String ={ "Tiny"}}

trait NonOvalFruit extends Fruit{ def getname():String ={ "Non Oval"}}

trait Berries extends OvalFruit with TinyFruit {override  def getname():String ={  super[TinyFruit].getname +" "+ super[OvalFruit].getname +" Berries"}}

trait Grapes extends TinyFruit with OvalFruit{ override def getname():String ={  super[TinyFruit].getname +" "+ super[OvalFruit].getname+" Grapes"}}

class Apple extends BigFruit with OvalFruit{
  override def getname():String ={
    super[BigFruit].getname +" "+ super[OvalFruit].getname + " Apple"
  }
}

class Orange extends BigFruit with OvalFruit{
   override def getname():String ={
    super[BigFruit].getname +" "+ super[OvalFruit].getname+ " Orange"
  }
}

class RedGrapes extends Grapes{
   override def getname():String ={
    super[Grapes].getname+" RedGrapes"
  }
}

class GreenGrapes extends Grapes{
   override def getname():String ={
     super[Grapes].getname+" GreenGrapes"
  }
}

class BlueBerry extends Berries{
   override def getname():String ={
     super[Berries].getname+" BlueBerry"
  }
}

class BlackBerry extends Berries{
   override def getname():String ={
    super[Berries].getname+" BlackBerry"
  }
}

class Rhubarb extends BigFruit with NonOvalFruit{
   override def getname():String ={
    super[BigFruit].getname +" "+ super[NonOvalFruit].getname+" Rhubarb"
  }
}


class BST {
  var root : Fruit =null
  var fruitList = List[Fruit]()
  
  def insert(node : Fruit){
    if (root eq null){
       root = node
    }
    else{
      this.insertcompare(root, node, compare)
    }
    
  }
  
 def insertcompare(root:Fruit ,node:Fruit , compare:(Fruit,Fruit)=>Boolean): Unit ={
   if(compare(root,node)==true){
     if(root.right==null){
       root.setright(node)
     }
     else{
       insertcompare(root.right,node,compare)
     }
   }
   else{
     if(root.left==null){
       root.setleft(node)
     }
     else{
       insertcompare(root.left,node,compare)
     }
   }
   
 }
  
  def compare (node1 : Fruit , node2:Fruit): Boolean={
    if(node2.weight>=node1.weight){
      return true;
    }
    else{
      return false;
    }
  }
  
  def Iterate(){
      fruitList=Nil
   iterator(root)
   fruitList = fruitList.reverse
   fruitList.foreach(fruitoflist => println("I am " +fruitoflist.getname()+ " and my weight is " + fruitoflist.weight))
    
  }
  
  def iterator(node : Fruit){
   if(node != null){
     iterator(node.left)
     fruitList = node :: fruitList
     iterator(node.right)
     
   }
    
  }
  
  def findheaviest(){
    var node = root
    while(node.right!=null){
      node = node.right
    }
     println("The heaviest fruit is " +node.getname()+ " and of weight " + node.weight)
  }
  
   def findLightest(){
    var node = root
    while(node.left!=null){
      node = node.left
    }
     println("The Lightest fruit is " +node.getname()+ " and of weight " + node.weight)
  }
   
  
  def matchtype(fruit:Fruit , typeoffruit : String)= typeoffruit match{
      case "Oval" => if(fruit.isInstanceOf[OvalFruit]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Tiny" => if(fruit.isInstanceOf[TinyFruit]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Big" => if(fruit.isInstanceOf[BigFruit]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Non Oval" => if(fruit.isInstanceOf[NonOvalFruit]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Berries" => if(fruit.isInstanceOf[Berries]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Grapes" => if(fruit.isInstanceOf[Grapes]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Apple" => if(fruit.isInstanceOf[Apple]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Orange" => if(fruit.isInstanceOf[Orange]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Red Grapes" => if(fruit.isInstanceOf[RedGrapes]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Green Grapes" => if(fruit.isInstanceOf[GreenGrapes]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Blue Berry" => if(fruit.isInstanceOf[BlueBerry]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Black Berry" => if(fruit.isInstanceOf[BlackBerry]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Rhubarb" => if(fruit.isInstanceOf[Rhubarb]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case "Fruit" => if(fruit.isInstanceOf[Fruit]) println("I am " +fruit.getname()+ " and my weight is " + fruit.weight)
      case _ =>

  }

   def filterbytype(typeoffruit : String){
       fruitList=Nil
       iterator(root)
       fruitList=fruitList.reverse
      
       fruitList.foreach(fruitoflist => matchtype(fruitoflist, typeoffruit) )
   }

   def filterbyweight(weight : Int){
       fruitList=Nil
       iterator(root)
       fruitList=fruitList.reverse
      
       fruitList.foreach(fruitoflist => if(fruitoflist.weight>=weight) println("I am " +fruitoflist.getname()+ " and my weight is " + fruitoflist.weight) )
   }

   def magnifyByType(typeoffruit: String , weight : Int){
       fruitList=Nil
       iterator(root)
       fruitList=fruitList.reverse
       root=null
       fruitList.foreach(fruitoflist =>typeoffruit match{
      case "Oval" => if(fruitoflist.isInstanceOf[OvalFruit]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null) ; insert(fruitoflist)
      case "Tiny" => if(fruitoflist.isInstanceOf[TinyFruit]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null); insert(fruitoflist)
      case "Big" => if(fruitoflist.isInstanceOf[BigFruit]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Non Oval" => if(fruitoflist.isInstanceOf[NonOvalFruit]) fruitoflist.weight=fruitoflist.weight+weight;fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Berries" => if(fruitoflist.isInstanceOf[Berries]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Grapes" => if(fruitoflist.isInstanceOf[Grapes]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Apple" => if(fruitoflist.isInstanceOf[Apple]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Orange" => if(fruitoflist.isInstanceOf[Orange]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Red Grapes" => if(fruitoflist.isInstanceOf[RedGrapes]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Green Grapes" => if(fruitoflist.isInstanceOf[GreenGrapes]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Blue Berry" => if(fruitoflist.isInstanceOf[BlueBerry]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Black Berry" => if(fruitoflist.isInstanceOf[BlackBerry]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Rhubarb" => if(fruitoflist.isInstanceOf[Rhubarb])  fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case "Fruit" => if(fruitoflist.isInstanceOf[Fruit]) fruitoflist.weight=fruitoflist.weight+weight; fruitoflist.setleft(null) ;fruitoflist.setright(null);insert(fruitoflist)
      case _ => fruitoflist.setleft(null) ;fruitoflist.setright(null); insert(fruitoflist)

  })
     

   }
  
}
object main {
   def main(args:Array[String]){  
      
    var t :BST= new BST()

    var a :Fruit= new Apple(); a.setweight(25)
    var b :BigFruit= new Orange(); b.setweight(20)
    var c :Fruit= new Orange(); c.setweight(36)
    var d :Berries= new BlueBerry(); d.setweight(10)
    var e :TinyFruit= new BlackBerry(); e.setweight(22)
    var f :OvalFruit= new Apple(); f.setweight(30)
    var g :Grapes= new RedGrapes(); g.setweight(40)
    var h :TinyFruit= new BlueBerry(); h.setweight(5)
    var i :Fruit= new Apple(); i.setweight(12)
    var j :Fruit= new Orange(); j.setweight(28)
    var k :Fruit= new GreenGrapes(); k.setweight(38)
    var l :Fruit= new Rhubarb(); l.setweight(48)

    t.insert(a);t.insert(b);t.insert(c);t.insert(d);t.insert(e);t.insert(f);t.insert(g);
    t.insert(h);t.insert(i);t.insert(j);t.insert(k);t.insert(l);

    t.Iterate();

    println(" ");

    t.findLightest();
    t.findheaviest();

     println("");
     println("These are the Berries");
    t.filterbytype("Berries")

     println("");
     println("These are the Apples");
    t.filterbytype("Apple")

     println("");
     println("These are Fruits greater than or equal 15");
    t.filterbyweight(15)

    println("");
     println("These are Fruits greater than or equal 40");
    t.filterbyweight(40)

    println("");
     println("These are Fruits after magnifing Tiny Fruits by 10 ");
    t.magnifyByType("Tiny", 10)
    t.Iterate()
    } 
}
